### PR TITLE
Refactor everything

### LINKED ISSUE #43: Refactor go-diff
Rework the repository code without breaking the API
- [x] Rename and move files around
- [x] Convert the existing tests to solely table-driven tests
- [x] Split the code to be more maintainable
- [x] Update the README to be more user and contributor friendly
- [x] Try to make the repository more contributor attractive
- [x] When done, ask for being reviewed and ping sergi that I am done here

### PR DESCRIPTION
**Pull request by zimmski**
This PR basically refactors the test code completely, the rest is moved around and sometimes refactored.

The goals are:

- Refactor the code so it is only one code style
- Split the code up into logical files
- Reuse as much code as possible. 
- Try to use the same variable names for the same functionality
- In general: Make the whole code more contributor friendly

It might be a good idea to look at each commit of the PR instead of the complete changes.

Fixes #43

### CODE REVIEW

**Comment by zimmski** [comment]
This is the coverage after refactoring the tests but before moving code around https://coveralls.io/builds/8655582 it is the same coverage as the current master.

**Comment by zimmski** [comment]
@sergi: It would be good if you could give me some feedback on this. I will refactor the rest of the code like in the last commit here 435c112fd793e8283659ae75013afd0009fcc6eb. I will basically refactor everything to be of one consistent code style, cleanup wherever I see something that could be done immediately (or add a "help wanted" issue) and basically make everything use Go idioms.

File: diffmatchpatch/dmp_test.go [hunk 1]
```diff
@@ -1531,8 +1531,8 @@ func Benchmark_DiffCommonSuffix(b *testing.B) {
 }
 
 func Benchmark_DiffMainLarge(b *testing.B) {
-	s1 := readFile("speedtest1.txt", b)
-	s2 := readFile("speedtest2.txt", b)
+	s1 := readFile("../testdata/speedtest1.txt", b)
```
Comment (marxriedl) [inline diffmatchpatch/dmp_test.go]: introduce constant for "../testdata", maybe even for speedtest1/2.txts

File: diffmatchpatch/diffmatchpatch_test.go [hunk 2]
```diff
@@ -1498,73 +1514,89 @@ func TestLastIndexOf(t *testing.T) {
...
 }
 
-func Benchmark_DiffMain(bench *testing.B) {
-	dmp := New()
-	dmp.DiffTimeout = time.Second
-	a := "`Twas brillig, and the slithy toves\nDid gyre and gimble in the wabe:\nAll mimsy were the borogoves,\nAnd the mome raths outgrabe.\n"
-	b := "I am the very model of a modern major general,\nI've information vegetable, animal, and mineral,\nI know the kings of England, and I quote the fights historical,\nFrom Marathon to Waterloo, in order categorical.\n"
+func BenchmarkDiffMain(bench *testing.B) {
+	s1 := "`Twas brillig, and the slithy toves\nDid gyre and gimble in the wabe:\nAll mimsy were the borogoves,\nAnd the mome raths outgrabe.\n"
+	s2 := "I am the very model of a modern major general,\nI've information vegetable, animal, and mineral,\nI know the kings of England, and I quote the fights historical,\nFrom Marathon to Waterloo, in order categorical.\n"
```
Comment (marxriedl) [inline diffmatchpatch/diffmatchpatch_test.go]: s1, s2 is not necessarily better than a, b

File: diffmatchpatch/diffmatchpatch_test.go [hunk 3]
```diff
@@ -1498,73 +1514,89 @@ func TestLastIndexOf(t *testing.T) {
...
+
 	// Increase the text lengths by 1024 times to ensure a timeout.
 	for x := 0; x < 10; x++ {
```
Comment (marxriedl) [inline diffmatchpatch/diffmatchpatch_test.go]: well, whatever this is and does i hope it does its job

File: diffmatchpatch/diffmatchpatch_test.go [hunk 4]
```diff
@@ -35,130 +39,134 @@ func pretty(diffs []Diff) string {
...
 	return s1, s2
 }
 
-func Test_diffCommonPrefix(t *testing.T) {
-	dmp := New()
-	// Detect any common suffix.
-	// Null case.
-	assert.Equal(t, 0, dmp.DiffCommonPrefix("abc", "xyz"), "'abc' and 'xyz' should not be equal")
+func TestDiffCommonPrefix(t *testing.T) {
+	type TestCase struct {
```
Comment (marxriedl) [inline diffmatchpatch/diffmatchpatch_test.go]: theres a lot of similar TestCase types, unify and extract

File: diffmatchpatch/diffmatchpatch_test.go [hunk 5]
```diff
@@ -810,214 +1006,303 @@ func Test_diffDelta(t *testing.T) {
...
+		actual := dmp.DiffMain(tc.Text1, tc.Text2, false)
+		assert.Equal(t, tc.Expected, actual, fmt.Sprintf("Test case #%d, %#v", i, tc))
+	}
+}
 
+func TestDiffMainWithTimeout(t *testing.T) {
+	dmp := New()
 	dmp.DiffTimeout = 200 * time.Millisecond
+
 	a := "`Twas brillig, and the slithy toves\nDid gyre and gimble in the wabe:\nAll mimsy were the borogoves,\nAnd the mome raths outgrabe.\n"
```
Comment (marxriedl) [inline diffmatchpatch/diffmatchpatch_test.go]: consider constant

File: diffmatchpatch/diffmatchpatch_test.go [hunk 6]
```diff
@@ -810,214 +1006,303 @@ func Test_diffDelta(t *testing.T) {
...
 	b := "I am the very model of a modern major general,\nI've information vegetable, animal, and mineral,\nI know the kings of England, and I quote the fights historical,\nFrom Marathon to Waterloo, in order categorical.\n"
```
Comment (marxriedl) [inline diffmatchpatch/diffmatchpatch_test.go]: consider constant

File: diffmatchpatch/diffmatchpatch_test.go (hunk 2 above)
Comment (zimmski) [inline diffmatchpatch/diffmatchpatch_test.go]: That is true, but the whole code base uses *1&*2 more often than just a&b

File: diffmatchpatch/diffmatchpatch_test.go (hunk 4 above)
Comment (zimmski) [inline diffmatchpatch/diffmatchpatch_test.go]: I prefer to not reuse the TestCase types. The reason is that that way I can keep them inside the test function. They are not exported and the way a test function is made stays always the same without the need to find the right TestCase type.

**Pull request merged**